File: src/main/org/nlogo/prim/threed/_breedat.java
package org.nlogo.prim.threed;

import org.nlogo.api.LogoException;
import org.nlogo.nvm.Reporter;
import org.nlogo.nvm.Syntax;

public final strictfp class _breedat
    extends Reporter {
  private final String breedName;

  public _breedat() {
    breedName = null;
  }

  public _breedat(String breedName) {
    this.breedName = breedName;
  }

  @Override
  public Syntax syntax() {
    int[] right = {Syntax.TYPE_NUMBER,
        Syntax.TYPE_NUMBER,
        Syntax.TYPE_NUMBER};
    int ret = Syntax.TYPE_TURTLESET;
    return Syntax.reporterSyntax(right, ret, "-TP-");
  }

  @Override
  public String toString() {
    return super.toString() + ":" + breedName;
  }

  @Override
  public Object report(final org.nlogo.nvm.Context context) throws LogoException {
    double dx = argEvalDoubleValue(context, 0);
    double dy = argEvalDoubleValue(context, 1);
    double dz = argEvalDoubleValue(context, 2);
    org.nlogo.agent.Patch patch = null;

    try {
      patch = ((org.nlogo.agent.Agent3D) (context.agent)).getPatchAtOffsets(dx, dy, dz);
    } catch (org.nlogo.api.AgentException e) {
      return new org.nlogo.agent.ArrayAgentSet(org.nlogo.agent.Turtle.class, 0,
          false, world);
    }


    if (patch == null) {
      return new org.nlogo.agent.ArrayAgentSet(org.nlogo.agent.Turtle.class, 0,
          false, world);
    }

    org.nlogo.agent.AgentSet agentset =
        new org.nlogo.agent.ArrayAgentSet(org.nlogo.agent.Turtle.class, patch.turtleCount(),
            false, world);
    org.nlogo.agent.AgentSet breed = world.getBreed(breedName);
    for (org.nlogo.agent.Turtle turtle : patch.turtlesHere()) {
      if (turtle != null && turtle.getBreed() == breed) {
        agentset.add(turtle);
      }
    }
    return agentset;
  }
}

File: netlogo-gui/src/main/prim/threed/_breedat.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim.threed

import org.nlogo.agent.{ Agent3D, AgentSet, AgentSetBuilder }
import org.nlogo.api.AgentException
import org.nlogo.core.AgentKind
import org.nlogo.nvm.{ Context, Reporter }

class _breedat(breedName: String) extends Reporter {
  def this() = this(null)

  override def toString =
    super.toString + ":" + breedName
  override def report(context: Context): AgentSet = {
    val dx = argEvalDoubleValue(context, 0)
    val dy = argEvalDoubleValue(context, 1)
    val dz = argEvalDoubleValue(context, 2)
    val patch =
      try context.agent.asInstanceOf[Agent3D].getPatchAtOffsets(dx, dy, dz)
      catch {
        case e: AgentException =>
          return AgentSet.emptyTurtleSet
      }
    if (patch == null)
      AgentSet.emptyTurtleSet
    else {
      val agentSetBuilder = new AgentSetBuilder(AgentKind.Turtle, patch.turtleCount)
      val breed = world.getBreed(breedName)
      val it = patch.turtlesHere.iterator
      while(it.hasNext) {
        val turtle = it.next()
        if (turtle != null && (turtle.getBreed eq breed))
          agentSetBuilder.add(turtle)
      }
      agentSetBuilder.build()
    }
  }
}
